runner/internal/backend/base/storage.go

```diff
@@ -0,0 +1,238 @@
+package base
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"github.com/dstackai/dstack/runner/internal/common"
+	"github.com/dstackai/dstack/runner/internal/gerrors"
+	"github.com/dstackai/dstack/runner/internal/log"
+	"golang.org/x/sync/errgroup"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type StorageObject struct {
+	// Key is relative to the queried prefix
+	Key     string
+	Size    int64
+	ModTime time.Time
+	Symlink string
+}
+
+type Storage interface {
+	Download(ctx context.Context, key string, writer io.Writer) error
+	Upload(ctx context.Context, reader io.Reader, key string) error
+	Delete(ctx context.Context, key string) error
+	Rename(ctx context.Context, oldKey, newKey string) error
+	CreateSymlink(ctx context.Context, key, symlink string) error
+	GetMetadata(ctx context.Context, key, tag string) (string, error)
+	// List generates StorageObject-s in lexicographical order
+	List(ctx context.Context, prefix string) (<-chan *StorageObject, <-chan error)
+}
+
+func (a StorageObject) Equals(b StorageObject) bool {
+	return a.Key == b.Key && a.Size == b.Size && a.ModTime == b.ModTime && a.Symlink == b.Symlink
+}
+
+func DownloadDir(ctx context.Context, storage Storage, key, dst string) error {
+	key = common.AddTrailingSlash(key)
+	dst = common.AddTrailingSlash(dst)
+	objects, errCh := storage.List(ctx, key)
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(TransferThreads)
+	for rObj := range objects {
+		obj := rObj // goroutines and range variables don't work together
+		objKey := path.Join(key, obj.Key)
+		filePath := filepath.Join(dst, obj.Key)
+		g.Go(func() error {
+			if err := os.MkdirAll(filepath.Dir(filePath), 0o755); err != nil {
+				return gerrors.Wrap(err)
+			}
+			if obj.Size == 0 && obj.Symlink != "" {
+				if err := os.Symlink(obj.Symlink, filePath); err != nil {
+					return gerrors.Wrap(err)
+				}
+			} else {
+				if err := DownloadFile(ctx, storage, objKey, filePath); err != nil {
+					return gerrors.Wrap(err)
+				}
+				if err := os.Chtimes(filePath, obj.ModTime, obj.ModTime); err != nil {
+					return gerrors.Wrap(err)
+				}
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return gerrors.Wrap(err)
+	}
+	select {
+	case err := <-errCh:
+		return gerrors.Wrap(err)
+	default:
+	}
+	return nil
+}
+
+// UploadDir save local files tree to the Storage
+// If delete is true — files existing only in storage would be deleted
+// If withoutChanges is true — files would always be uploaded despite the same size and modification time
+func UploadDir(ctx context.Context, storage Storage, src, key string, delete, withoutChanges bool) error {
```

Comment:
Can you refactor this func into subfunctions? Hard to grasp what's going on.

Reply:
Sure. I will refactor it within the next PR